### ISSUE TITLE
[Bug] AcquireTokenSilent should not make calls to the network unless required

### ISSUE THREAD

**Issue by tomeverin** [AzureAD/microsoft-authentication-library-for-dotnet#1566]
**Which Version of MSAL are you using ?**
MSAL 4.7.1

**Platform**
Xamarin.Android

**What authentication flow has the issue?**
* Desktop / Mobile
    * [x] Interactive
    * [ ] Integrated Windows Auth
    * [ ] Username Password
    * [ ] Device code flow (browserless)
* Web App
    * [ ] Authorization code
    * [ ] OBO
* Web API
    * [ ] OBO

**Is this a new or existing app?**
This is a new app

**Repro**

```csharp
IEnumerable<IAccount> accounts = await PCA.GetAccountsAsync();
IAccount account = GetAccountByPolicy(accounts, B2CConstants.PolicySignUpSignIn);

AuthenticationResult ar = await PCA.AcquireTokenSilent(B2CConstants.Scopes, account)
                    .WithB2CAuthority(B2CConstants.Authority)
                    .ExecuteAsync();
```

**Expected behavior**
If there is a valid token in the TokenCache, return Authentication Result without network call; fallback refresh token by network call only if required.

**Actual behavior**
If there is a valid token in the TokenCache, and no network connectivity (android emulator, flight mode), a network call is attempted which leads to an exception: 

> Java.Net.UnknownHostException: 'Unable to resolve host "xxx.b2clogin.com": No address associated with hostname'

**Additional context/ Logs / Screenshots**
Have confirmed that the returned account is correct. Can test by turning flight mode on and off; the above always fails with the UnknownHostException when in flight mode; even moments after successful silent login with connectivity enabled.

**Possible Solution**
I saw in https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1174 :

> GetAccountsAsync and AcquireTokenSilent (when a valid AT is in the cache) no longer make network calls unless an atypical authority host is used (i.e. something other than https://login.microsoft.com etc.). For a full list of these authorities, see https://aka.ms/aad-instance-discovery

b2clogin.com is now offered as a typical authority host, this similarly should not lead to a network call?

**Comment by bgavrilMS** [comment]
What scopes do you ask for?

**Comment by tomeverin** [comment]
> 
> 
> What scopes do you ask for?

Scopes string is empty = ""

**Comment by bgavrilMS** [comment]
This is sort of expected because the scope "" is not a valid scope. MSAL should throw in this case, but it does not.

MSAL is a library that is used to request access tokens, i.e. tokens that can be used to access a protected API. What it looks like you are trying to do is to just log-in the user, i.e. you want just an IDToken, which provides things like an email and a username. This is not currently supported by MSAL, but it is on our roadmap (@jmprieur - perhaps prioritize this for 4.9?)https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/715

A workaround might be to ask for the "email" scope - see if that works. 

OAuth spec states that the token issuer MUST return a token for all the valid scopes. So internally MSAL asks a token for scopes "profile" "openid" "offline_access" and "". ESTS response with a token for "openid" and "profile" . So replaying the original request results in a cache miss.

**Comment by tomeverin** [comment]
Yes, I am trying to achieve log-in only. I'd really like for the offline scenario to work if a valid token is available in the cache. I'm using B2C for local account only, and am not interested in the social identity provider account support offered by B2C.

Passing scopes = { "email" } made no difference. I also tried setting up a scope in B2C portal, and passed scopes = { "https://{xxx}.onmicrosoft.com/{xxx}/User.Read" }, but this made not difference either.

Is the fact B2C doesn't return `preferred_username` in the token still relevant to MSAL caching? The returned account is the right one with correct `ObjectID` and `TenantID`, but has `username: "Missing from the token response"`.

From https://github.com/Azure-Samples/active-directory-b2c-dotnetcore-webapp/issues/64 I see the following:

> This is a known issue w/B2C. The reason is because B2C does not return a value in the IdToken for preferred username because of the limitations with the social accounts. For example, AAD does return a value here because it knows who the user is, but for b2c, because the user can sign in with a local account, FB, Github, etc...there is not a consistent value for them to use for preferred username. Also, some social accounts (like Github) don't require email. So...to unblock MSAL from rolling out cache compatibility with ADAL, we decided to use "Missing from the token response" on our end when dealing with the B2C accounts.

However the following (https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-net-aad-b2c-considerations) suggests that finding account by policy is enough to mitigate the problem:
> Known issue with Azure AD B2C:  MSAL.Net supports a token cache. The token caching key is based on the claims returned by the Identity Provider. Currently MSAL.Net needs two claims to build a token cache key:
> 
>     tid which is the Azure AD Tenant ID, and
>     preferred_username
> 
> Both these claims are missing in many of the Azure AD B2C scenarios.
> 
> ... Azure AD B2C does not return a value in the IdToken for the preferred_username because of limitations with the social accounts and external identity providers ... MSAL must return a value for preferred_username to maintain cache compatibility across libraries.
> Workarounds - Mitigation for the missing tenant ID
> 
> The suggested workaround is to use the Caching by Policy
> 
> Alternatively, you can use the tid claim, if you are using the B2C custom policies, because it provides the capability to return additional claims to the application

**Comment by bgavrilMS** [comment]
@TiagoBrenck - can you help ?

**Comment by TiagoBrenck** [comment]
@bgavrilMS I will help investigate this error.

@tomeverin, about the `preferred_username` question, if you could get the correct `IAccount` filtering by policy, then you are already good to go. The `IAccount` object will indeed return `username: Missing from the token response` but this is irrelevant for B2C caching. It blocks you from using logics like

```
var accounts = await application.GetAccountsAsync().ConfigureAwait(false);
account = accounts.FirstOrDefault(a => a.Username == loggedUserLogin);
```
but for B2C, you shouldn't use it anyways because the user vs IAccount relationship is 1-to-many (having policyId as the unique field).

I will investigate the network error and keep you posted with further news.

**Comment by TiagoBrenck** [comment]
@tomeverin I couldn't repro the error. I have used https://github.com/Azure-Samples/active-directory-b2c-xamarin-native as code sample and performed the following steps (using emulator and real Android device):

1. With the network on, signed in for the first time (typed username and password)
1. Turned off the network and turned on airplane mode
1. Killed the running app activity on Android
1. Opened the app again and pressed *Sign In* button
1. I got signed in automatically

Am I missing anything?

**Comment by tomeverin** [comment]
Now working as expected when passing a valid scope ✔

Thanks for having a look @TiagoBrenck . I compared the sample code with my code, and couldn't find any functional differences, except the sample has an API scope when acquiring tokens... as @bgavrilMS had discussed above.

I found a different way in the B2C tenant portal to add a scope to the application, and have got MSAL working as expected. Glad that https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/715 is on the roadmap for this login-only use case.

---
For others attempting to configure B2C to get this working, it works with the User.Read test scope as follows:

1. Enter B2C Tenant
2. Go into Applications -> Choose relevant application -> "API Access" sidebar item -> "+ Add" -> 
"User Read Test (User.Read)" -> Save
3. Go back to the B2C Tenant
4. Go into "App Registrations (Preview)" sidebar item -> Choose relevant application -> "API Permissions" sidebar item -> Click the "User.Read" permission -> Copy the URL from the modal blade that appears
5. The URL from 4. is the scope

The B2C Tenant application does not need to include web app / web api for this to work.

I have described the steps in this way because it was not initially clear to me what the correct URL would be for my User.Read scope; the B2C "Application Registrations" view which is in preview makes this more obvious.
